haskell/src/Lorentz/Contracts/Spec/TZIP16Interface.hs

```diff
@@ -0,0 +1,241 @@
+data MichelsonStorageView st param ret = MichelsonStorageView
+  { msvCode :: Instr '[ 'TPair st param] '[ret]
+  , msvAnnotations :: [(SomeAnn, Text)]
+  }
+
+----------------------------------------------------------------------------
+-- JSON serializers/deserializers
+----------------------------------------------------------------------------
+
+deriveJSON aesonOptions ''License
+deriveJSON aesonOptions ''Source
+deriveJSON aesonOptions ''Error
+instance KnownT st => ToJSON (Metadata st) where toJSON = genericToJSON aesonOptions
+instance KnownT st => FromJSON (Metadata st)  where parseJSON = genericParseJSON aesonOptions
+instance KnownT st => ToJSON (View st)  where toJSON = genericToJSON aesonOptions
+instance KnownT st => FromJSON (View st)  where parseJSON = genericParseJSON aesonOptions
+
+instance KnownT st => ToJSON (ViewImplementation st) where
+  toJSON (VIMichelsonStorageView smsv) =
+    object [ "michelson-storage-view" .= smsv ]
+
+instance KnownT st => FromJSON (ViewImplementation st) where
+  parseJSON = withObject "ViewImplementation" $ \obj -> do
+    smsv <- obj .: "michelson-storage-view"
+    pure (VIMichelsonStorageView smsv)
+
+
+instance ToJSON (SomeMichelsonStorageView st) where
+  toJSON (SomeMichelsonStorageView msv) = toJSON msv
+
+instance (KnownT st, KnownT param, KnownT ret) => ToJSON (MichelsonStorageView st param ret) where
+  toJSON (MichelsonStorageView code anns) =
+    object
+      [ "parameter" .= toExpression @T (demote @param)
+      , "return-type" .= toExpression @T (demote @ret)
+      , "code" .= toExpression code
+      , "annotations" .= annsJSON
+      ]
+    where
+      annsJSON = anns <&> \(someAnn, description) ->
+        object
+          [ "name" .= someAnn
+          , "description" .= description
+          ]
+
+instance KnownT st => FromJSON (SomeMichelsonStorageView st) where
+  parseJSON = withObject "MichelsonStorageView" $ \obj -> do
+
+    annotations <- obj .: "annotations"
+    paramT <- obj .: "parameter" >>= evalEither . fromExpression @T
+    returnTypeT <- obj .: "return-type" >>= evalEither . fromExpression @T
+
+    withSomeSingT paramT $ \(_ :: Sing param) ->
+      withSomeSingT returnTypeT $ \(_ :: Sing ret) -> do
+        instr <- obj .: "code" >>= evalEither . fromExpression @(Instr '[ 'TPair st param] '[ret])
+        pure $
+          SomeMichelsonStorageView @st
+            (MichelsonStorageView @st @param @ret instr annotations)
+    where
+      evalEither :: Either FromExpressionError a -> Parser a
+      evalEither (Right a) = pure a
+      evalEither (Left err) = fail (pretty err)
+
+----------------------------------------------------------------------------
+-- Handling views
+----------------------------------------------------------------------------
+
+mkMichelsonStorageView :: [(SomeAnn, Text)] -> Lambda (st, param) ret -> MichelsonStorageView (ToT st) (ToT param) (ToT ret)
+mkMichelsonStorageView anns code =
```

Comment:
Some instructions seems to be forbidden in the view code. Should we check that here?

Reply:
Which instructions? :open_mouth:

Reply:
I quite probably have mixed something up lol, but here is what I was referring to

https://gitlab.com/tzip/tzip/-/blob/eb1da57684599a266334a73babd7ba82dbbbce66/proposals/tzip-16/tzip-16.md#michelson-storage-views

```
There are limitations on the contents of the "code" field:

The following instructions must not be used:
AMOUNT,
CREATE_CONTRACT,
SENDER,
SET_DELEGATE,
SOURCE,
and
TRANSFER_TOKENS
```

Reply:
Ah good catch, I missed that completely 😅 

This is not a problem in this specific instance, because we know these views do not use those instructions, so I propose not doing it now, and add it as an item to this issue: https://gitlab.com/morley-framework/morley-metadata/-/issues/2

What do you think @gromakovsky ?

Reply:
Yes, sounds good to me.

Reply:
> add it as an item to this issue: https://gitlab.com/morley-framework/morley-metadata/-/issues/2

So can you please add a comment to that issue or just update its description?

Reply:
> propose not doing it now, and add it as an item to this issue

Fine by me.